### PR TITLE
API Add history viewer React components

### PR DESCRIPTION
**Pull request by robbieaverill**
Parent issue: https://github.com/silverstripe/silverstripe-versioned/issues/37

### CODE REVIEW

**Comment by raissanorth** [comment]
There's a CMS menu item that shouldn't be there:

image

**Comment by robbieaverill** [comment]
Another issue: the injector is not aware of HistoryViewer on initial load, however refreshing the page works. This is reproducible by navigating out to campaigns for example then back into a page history tab via Pages section.

File: src/Controllers/CMSPageHistoryViewerController.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+use SilverStripe\CMS\Controllers\CMSMain;
+use SilverStripe\Forms\FieldList;
+use SilverStripe\Forms\Form;
+use SilverStripe\ORM\DataObject;
+use SilverStripe\VersionedAdmin\Forms\HistoryViewerField;
+
+if (!class_exists(CMSMain::class)) {
+    return;
+}
```
[the remark this answers is not in the document]
Comment (robbieaverill) [reply src/Controllers/CMSPageHistoryViewerController.php]: I'm not sure :) We could remove it for now anyway. Thanks for all the feedback, I'll follow up tomorrow

File: client/src/components/HistoryViewer/HistoryViewerVersionState.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  /**
+   * Formats the last edited date according to the current locale and return it in the example
+   * format "03/01/2018 2:45 PM"
+   *
+   * @returns {string}
+   */
+  getDate() {
+    moment.locale(i18n.detectLocale());
+    return moment(this.props.version.LastEdited).format('L LT');
+  }
```
[the remark this answers is not in the document]
Comment (robbieaverill) [reply client/src/components/HistoryViewer/HistoryViewerVersionState.js]: How stuck are you on this? Looks like moment supports more locales than date-fns does (at the moment). Do you think we could log this as an enhancement for later?

[the remark this answers is not in the document]
Comment (unclecheese) [reply client/src/components/HistoryViewer/HistoryViewerVersionState.js]: Not very. Just noting it. If there's that big a gap in locale support, that's a good enough reason to go with moment for now, but just noting that it is a relatively expensive dependency given the size of the API surface we actually use.

[the remark this answers is not in the document]
Comment (robbieaverill) [reply client/src/components/HistoryViewer/HistoryViewerVersionState.js]: Fair call. We could replace it in silverstripe/admin at the same time at a later date, perhaps

File: src/Controllers/CMSPageHistoryViewerController.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (unclecheese) [reply src/Controllers/CMSPageHistoryViewerController.php]: Yup, just confirmed that's correct. We're doing it in other places, too, like the graphql classes in `versioned`. I didn't realise everything got loaded regardless of the implementation of PSR-4. So leave this one.

File: client/src/components/HistoryViewer/HistoryViewerVersionList.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import React, { PureComponent } from 'react';
+import HistoryViewerHeading from 'components/HistoryViewer/HistoryViewerHeading';
+import HistoryViewerVersion from 'components/HistoryViewer/HistoryViewerVersion';
+
+class HistoryViewerVersionList extends PureComponent {
+  render() {
```
[the remark this answers is not in the document]
Comment (robbieaverill) [reply client/src/components/HistoryViewer/HistoryViewerVersionList.js]: jest has problems testing it, so I've used a PureComponent instead. Another example of this causing problems: https://github.com/silverstripe/silverstripe-admin/pull/406

**Comment by robbieaverill** [comment]
Ok @unclecheese I've pushed an update with most of your suggestions implemented! Thanks =)

File: client/src/components/HistoryViewer/HistoryViewerVersionState.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (robbieaverill) [reply client/src/components/HistoryViewer/HistoryViewerVersionState.js]: https://github.com/silverstripe/silverstripe-versioned-admin/issues/4

**Comment by robbieaverill** [comment]
Self merging with @unclecheese's approval

**Pull request merged**